vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_fetcher.go

```diff
@@ -207,25 +208,85 @@ func (f *controllerFetcher) getParentOfController(controllerKey ControllerKeyWit
 		return getParentOfWellKnownController(informer, controllerKey)
 	}
 
-	// TODO: cache response
-	groupVersion, err := schema.ParseGroupVersion(controllerKey.ApiVersion)
+	groupKind, err := controllerKey.groupKind()
 	if err != nil {
 		return nil, err
 	}
-	groupKind := schema.GroupKind{
-		Group: groupVersion.Group,
-		Kind:  controllerKey.Kind,
-	}
 
 	owner, err := f.getOwnerForScaleResource(groupKind, controllerKey.Namespace, controllerKey.Name)
-	if err != nil {
+	if apierrors.IsNotFound(err) {
+		return nil, nil
+	} else if err != nil {
 		return nil, fmt.Errorf("Unhandled targetRef %s / %s / %s, last error %v",
 			controllerKey.ApiVersion, controllerKey.Kind, controllerKey.Name, err)
 	}
 
 	return owner, nil
 }
 
+func (c *ControllerKeyWithAPIVersion) groupKind() (schema.GroupKind, error) {
+	// TODO: cache response
+	groupVersion, err := schema.ParseGroupVersion(c.ApiVersion)
+	if err != nil {
+		return schema.GroupKind{}, err
+	}
+
+	groupKind := schema.GroupKind{
+		Group: groupVersion.Group,
+		Kind:  c.ControllerKey.Kind,
+	}
+
+	return groupKind, nil
+}
+
+func (f *controllerFetcher) isWellKnown(key *ControllerKeyWithAPIVersion) bool {
+	kind := wellKnownController(key.ControllerKey.Kind)
+	_, exists := f.informersMap[kind]
+	return exists
+}
+
+func (f *controllerFetcher) isScalable(key *ControllerKeyWithAPIVersion) (bool, error) {
+	groupKind, err := key.groupKind()
+	if err != nil {
+		return false, err
+	}
+
+	mappings, err := f.mapper.RESTMappings(groupKind)
+	if err != nil {
+		return false, err
+	}
+	var lastError error
+	for _, mapping := range mappings {
+		groupResource := mapping.Resource.GroupResource()
+		_, err := f.scaleNamespacer.Scales(key.Namespace).Get(groupResource, key.Name)
```

Comment:
This code is very similar to getOwnerForScaleResource, and we are performing a Get call to the API Server here if I'm reading this right and that is expensive. We should make sure that for each controllerKey we only perform this once while traversing the owner ref chain (Currenlty it may be twice, to check the owner and then check isScalable).

Reply:
@bskiba, so we've changed the code to only call this once, I've done this by setting a `scalable` boolean field on the controllers so we know if they implement the scale sub-resource and can re-use the information. Let me know if you think this is better.